### PR TITLE
Fix logic on handling body requests

### PR DESCRIPTION
**Pull request by jhs7jhs**

### CODE REVIEW

**Comment by jhs7jhs** [comment]
No, I found a possible race condition scenario now. Still need to fix.

File: sync/src/block/extension.rs [hunk 1]
```diff
@@ -182,28 +203,34 @@ impl NetworkExtension for Extension {
...
-                        let expired_request = request_list.iter().find(|(r, _)| *r == request_id).cloned();
-                        if let Some((request_id, request)) = expired_request {
-                            match request {
-                                RequestMessage::Bodies(hashes) => {
-                                    self.body_downloader.lock().reset_downloading(&hashes);
-                                }
-                                _ => {}
+                let mut tokens_info = self.tokens_info.write();
+                let token_info = tokens_info.get_mut(&token).unwrap();
+                if token_info.1 == None {
```
Comment (sgkim126) [inline sync/src/block/extension.rs]: token_info.1.is_none()

File: sync/src/block/extension.rs [hunk 2]
```diff
@@ -182,28 +203,34 @@ impl NetworkExtension for Extension {
...
+                    return
+                }
+
+                let id = token_info.0;
+                let request_id = token_info.1.unwrap();
+                if let Some(request_list) = requests.get(&id) {
```
Comment (sgkim126) [inline sync/src/block/extension.rs]: requests.get(&id).expect("...")

File: sync/src/block/extension.rs [hunk 3]
```diff
@@ -288,10 +315,18 @@ impl Extension {
...
+        let mut tokens_info = self.tokens_info.write();
         if peers.contains_key(from) {
             peers.get_mut(from).unwrap().update(total_score, best_hash);
         } else {
             requests.insert(*from, Vec::new());
+            if let Some(token) = self.token_generator.lock().gen() {
+                tokens_info.insert(token, (*from, None));
+                tokens.insert(*from, token);
+            } else {
+                cinfo!(SYNC, "Token generator is full");
```
Comment (sgkim126) [inline sync/src/block/extension.rs]: cerror

File: sync/src/block/extension.rs [hunk 4]
```diff
@@ -393,16 +428,32 @@ impl Extension {
...
-                    self.on_body_response(hashes, bodies)
+                    if let Some(token) = self.tokens.read().get(from) {
+                        if let Some(token_info) = self.tokens_info.write().get_mut(token) {
+                            if token_info.1 == None {
+                                ctrace!(SYNC, "Expired before handling response");
+                                return
+                            }
+                            self.api.lock().as_ref().map(|api| {
+                                api.clear_timer(*token)
+                                    .or(ctrace!(SYNC, "Timer is already called, but response handling is faster"));
```
Comment (sgkim126) [inline sync/src/block/extension.rs]: .or_else(|| ctrace(...))

File: sync/src/block/extension.rs [hunk 5]
```diff
@@ -393,16 +428,32 @@ impl Extension {
...
-                    self.on_body_response(hashes, bodies)
+                    if let Some(token) = self.tokens.read().get(from) {
+                        if let Some(token_info) = self.tokens_info.write().get_mut(token) {
+                            if token_info.1.is_none(){
+                                ctrace!(SYNC, "Expired before handling response");
+                                return
+                            }
+                            self.api.lock().as_ref().map(|api| {
+                                api.clear_timer(*token)
+                                    .or_else(|_| ctrace!(SYNC, "Timer is already called, but response handling is faster"));
```
Comment (sgkim126) [inline sync/src/block/extension.rs]:
```rust
.or_else(|e| {
    ctrace!(SYNC, "Timer is already called, but response handling is faster");
    e
});
```

File: sync/src/block/extension.rs [hunk 6]
```diff
@@ -288,10 +315,18 @@ impl Extension {
 
         let mut requests = self.requests.write();
         let mut peers = self.header_downloaders.write();
+        let mut tokens = self.tokens.write();
+        let mut tokens_info = self.tokens_info.write();
         if peers.contains_key(from) {
             peers.get_mut(from).unwrap().update(total_score, best_hash);
         } else {
             requests.insert(*from, Vec::new());
+            if let Some(token) = self.token_generator.lock().gen() {
```
Comment (sgkim126) [inline sync/src/block/extension.rs]: I think it must be `self.token_generator.lock().gen().expect(...)`.

**Pull request merged**